File: sorting/bubble_sort.rb
def swap(arr, idx, idx_n)
  tmp = arr[idx]
  arr[idx] = arr[idx_n]
  arr[idx_n] = tmp
end

def bubble_sort_iteration(arr)
  num_swaps = 0
  arr.each_with_index do |n, idx|
    next if idx + 1 == arr.count
    next_n = arr[idx + 1]
    if n > next_n
      swap(arr, idx, idx + 1)
      num_swaps += 1
    end
  end
  num_swaps
end

def bubble_sort(arr)
  loop until bubble_sort_iteration(arr) == 0
  arr
end

puts bubble_sort [6, 1, 9, 15, 1, 2, 7, 3, 19, 2, 17]

File: sorting/bubble_sort.py
def swap(a, i, j):
    a[i], a[j] = a[j], a[i]

a = [1, 3, 4, 2, 8, 4, 5, 1, 10, 6, 8, 9, 25, 15, 13, 12, 18, 17, 3, 5, 2, 7]

def _bubble_sort(a):
    num_iterations = 0
    for i, n in enumerate(a):
        if i == len(a) - 1:
            continue
        if a[i] > a[i + 1]:
            num_iterations += 1
            swap(a, i, i + 1)

    return num_iterations

def bubble_sort(a):
    while True:
        num_iterations = _bubble_sort(a)
        if num_iterations == 0:
            break

print(a)
bubble_sort(a)
print(a)
